KAFKA-7197 expand gradle build: include Scala 2.13

dejan2609 opened the pull request:
**_Final PR description (June 2019):_**
- include Scala 2.13 into gradle build
- handle future milestone and RC versions of Scala in a better way
- if no Scala version is specified, default to scala 2.12 (bump from 2.11)
- include certain Xlint options (removed by Scala 2.13) for Scala 2.11/2.12 build only
- upgrade versions for dependencies: 
  - scalaLogging: 3.9.0 -->> 3.9.2
  - scalatest:        3.0.7 -->> 3.0.8
  - scoverage:       1.3.1 -->> 1.4.0

 - - - -

**_Initial PR description (August 2018):_**

details:
  * build expanded in order to use Scala 2.13 milestone version (version 2.13.0-M3 is used)
  * 'scoverage' version upgraded: 1.3.1 -->> 1.4.0-M3

Note: _**./gradlew -PscalaVersion=2.13 jar -q**_ returns few errors (well, as expected):

```
8 errors found

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':core:compileScala'.
> Compilation failed
```

Upgrade to Scala 2.13.0-M4 depends on: 
- https://github.com/scoverage/scalac-scoverage-plugin/pull/229  **_Add Scala 2.13.0-M4 support_**

ijuma commented:
Thanks for the PR. M5 is due soon, I think we should go straight to that in preparation of RC1. We should probably not build it by default until the final release.

ijuma commented on gradle/dependencies.gradle:
```diff
@@ -22,29 +22,37 @@ ext {
   
   // Enabled by default when commands like `testAll` are invoked
-  defaultScalaVersions = [ '2.11', '2.12' ]
+  defaultScalaVersions = [ '2.11', '2.12', '2.13' ]
```
I suggest we add this line after 2.13 is released.

dejan2609 replied:
Sure @ijuma, I agree on all accounts. 

Scala **2.13.0-M5** will be stable enough; (a way to early) PR was inspired by this JIRA ticket: https://issues.apache.org/jira/browse/KAFKA-7197 ***Release a milestone build for Scala 2.13.0 M3***

dejan2609 replied:
Resolving conversation (see git commit changes).

ijuma commented:
@dejan2609 2.13.0-M5 has been released. Maybe update this PR to use that?

dejan2609 commented:
@ijuma Sure, sorry for my late answer (I was traveling). I will try to update PR asap.

dejan2609 commented:
Update / related PR: 
https://github.com/scoverage/scalac-scoverage-plugin/issues/234 **_Add Scala 2.13.0-M5 support_**

dejan2609 commented:
Update 2: related PR mentioned above is solved, I will try to accommodate this PR accordingly.

dejan2609 commented:
Long overdue, I will try to squeeze this in a next few days.
FYI @ijuma

dejan2609 commented:
Ok, I have some solution, but I need to hack Xlint options in addition, suppress unused import statements temporarily and so on  (after that I will push some solution here):
```
dejans@BEG-IT-09 MINGW64 /c/Work/IDEA-2019.1/workspace/kafka (trunk)
$ ./gradlew -x test -PscalaVersion=2.13 jar -q
Building project 'core' with Scala version 2.13.0-M5
Building project 'streams-scala' with Scala version 2.13.0-M5
'by-name-right-associative' is not a valid choice for '-Xlint'
'unsound-match' is not a valid choice for '-Xlint'
<=------------> 14% EXECUTING [14s]
> :core:compileScala

dejans@BEG-IT-09 MINGW64 /c/Work/IDEA-2019.1/workspace/kafka (trunk)
$ ./gradlew -x test -PscalaVersion=2.12 jar -q
Building project 'core' with Scala version 2.12.8
Building project 'streams-scala' with Scala version 2.12.8
<=------------> 14% EXECUTING [9s]
> :core:compileScala

dejans@BEG-IT-09 MINGW64 /c/Work/IDEA-2019.1/workspace/kafka (trunk)
$ date
Tue Feb 26 08:25:43 CEST 2019

dejans@BEG-IT-09 MINGW64 /c/Work/IDEA-2019.1/workspace/kafka (trunk)

```

dejan2609 committed:
KAFKA-7197 gradle build changes: include Scala 2.13 into available and default Scala versions

details/notes:
  * gradle build now includes Scala 2.13; Scala 2.11/2.12 builds are intact
  * if no Scala version is specified, default to scala 2.12 (was 2.11)
  * 'baseScala' version resolving is expanded in order to handle milestone and RC versions of Scala
  * Scala 2.13 removes some scalac options and hence gradle build will include certain Xlint options ('by-name-right-associative' and 'unsound-match') for Scala 2.11/2.12 only
  * 'README.md' is changed accordingly
  * required version upgrades for dependencies:
        ** scalaLogging: 3.9.0 -->> 3.9.2
        ** scalatest:    3.0.7 -->> 3.0.8
        ** scoverage:    1.3.1 -->> 1.4.0

dejan2609 commented:
Ok, I pushed commit (after quite a while). 
Build passes locally for Scala 2.11 and 2.12 and (unsurprisingly) fails for Scala 2.13 with bunch of errors.

@ijuma now I see that I forgot to comply with your comment above (2.13 should not be listed among **_defaultScalaVersions_** yet... I will fix that).

dejan2609 commented:
Just to sum it up: 
- 2.13 is moved from **_defaultScalaVersions_** to **_availableScalaVersions_**
- only 1 test failed (not related with these changes, maybe test is flaky)
- now I see that I could check for a **dash** instead for a letter **M**, i.e. 
 `versions.scala.contains('-')` is much better than `versions.scala.contains('M')` because that way both milestones and release candidates would be covered:
-- https://github.com/scala/scala/releases/tag/v2.13.0-M5
-- https://contributors.scala-lang.org/t/2-13-0-rc1-timing-january-2019/2375

Let me know what you think @ijuma @ewencp

dejan2609 commented:
retest this please

dejan2609 commented:
retest this please

dejan2609 commented:
@ijuma please review (note: there are some flaky test).

dejan2609 commented:
@ijuma please review.

ewencp commented on build.gradle:
```diff
@@ -423,6 +421,13 @@ subprojects {
       ]
     }
 
+    if ( !versions.scala.startsWith('2.13.0-') ) {
```
do we think these are permanent? if they potentially aren't, i think it's better to leave in the default list (and git history) and remove based on the opposite conditional here.

ewencp replied:
hmm, my point still holds but I see now we also use this pattern above. In both cases I guess part of the problem is also that we don't have good version comparisons, just strings -- it's unfortunate that the previous one couldn't be `> 2.11` (and same here but opposite, `<= 2.12`).

dejan2609 replied:
Correct, my intention was to change as little as possible, so I just mirrored/applied existing comparison. Maybe it is possible to come up with a better solution, I will have to digg deeper.

dejan2609 replied:
> do we think these are permanent? if they potentially aren't, i think it's better to leave in the default list (and git history) and remove based on the opposite conditional here.

Done. I decided for a `if ( versions.baseScala.startsWith('2.13') )` (at least for now).

ewencp commented on gradle/dependencies.gradle:
```diff
@@ -82,7 +89,7 @@ versions += [
   rocksDB: "5.15.10",
   scalatest: "3.0.5",
-  scoverage: "1.3.1",
+  scoverage: "1.4.0-M5",
```
If we're going to depend on pre-release versions because of the scala version, we should probably leave a comment by the scala version with a note of everywhere that should be updated with it.

dejan2609 replied:
Agree, one should be skeptic about merging pre-release dependency version into a trunk, options: 
 1. suggested by a reviewer: leave appropriate comments and merge into trunk (this seems like a clean/easy solution)
 2. optional solution: merge into a separate (2.13 related) branch maybe ?

I will try number  _**(1).**_

dejan2609 replied:
Done, I left related comment(s) that could be seen as an overzealous/overkill (will make them short if required or move them into commit message).

ewencp commented on gradle/dependencies.gradle:
```diff
@@ -25,26 +25,33 @@ ext {
+if ( !versions.scala.startsWith('2.13.0-') ) {
+  versions["baseScala"] = versions.scala.substring(0, versions.scala.lastIndexOf("."))
+} else {
+  versions["baseScala"] = versions.scala
```
Why does this need to be different? Is this just to make it clear it's not a valid final release with that version of scala or does something not work otherwise?

If we really do need this, you can make it generic by just checking for `-` in the Scala version string and we won't have to edit it either a) when we update to final 2.13.0 or b) if we want to test against future scala version milestones, i.e. 2.14+.

dejan2609 replied:
@ewencp Makes sense. Actually, checking for a dash `-` only was one of interim solutions (now rebased).

dejan2609 replied:
> Why does this need to be different? Is this just to make it clear it's not a valid final release with that version of scala or does something not work otherwise?

  One word: cross-compiling. This hack (so to say) is added with intention to resolve **_scoverage-plugin_**, **_scoverage-runtime_**, **_scala-logging_** and **_scalatest_**  artifacts versions builded against various `2.13.0-` Scala available versions (and also to preserve compiling for a current default versions, i.e. for Scala 2.11 and 2.12). 

```
  scalaLogging: "com.typesafe.scala-logging:scala-logging_$versions.baseScala:$versions.scalaLogging",
  scalatest: "org.scalatest:scalatest_$versions.baseScala:$versions.scalatest",
  scoveragePlugin: "org.scoverage:scalac-scoverage-plugin_$versions.baseScala:$versions.scoverage",
  scoverageRuntime: "org.scoverage:scalac-scoverage-runtime_$versions.baseScala:$versions.scoverage",
```
(https://search.maven.org/search?q=g:org.scoverage%20a:scalac-scoverage-runtime_2.1*)

![image](https://user-images.githubusercontent.com/19467899/54521859-1fc7e380-496c-11e9-80b9-dcdae4a4cb86.png)

dejan2609 replied:
Dash `-` checking added via new commit (also, new commit comments explain this, I hope). 
I will leave this review open (Maven central snapshot above should be visible, it explains a lot).

ijuma replied:
@dejan2609 I think you should mention in the comment that pre-releases are not binary compatible with each other and that's the reason why libraries include the full Scala release string in their name for pre-releases. Focusing on cross-compilation doesn't explain the motivation and it's more of an implementation detail.

dejan2609 replied:
Done.

dejan2609 commented:
Thanx @ewencp for your review(s) above, I will try to implement/modify based on your suggestions.

dejan2609 commented:
New commit pushed. @ewencp  and @ijuma: feel free to check and review.

I added related comment(s) that could be seen as an overzealous/overkill (will make them short if required or move them into commit message).

dejan2609 commented:
Update: Scala **2.13.0-RC1** has just been released:
- https://github.com/scala/scala/releases/tag/v2.13.0-RC1
- https://contributors.scala-lang.org/t/2-13-0-rc1-timing-january-2019/2375/11

dejan2609 commented:
Please note that now Gradle can be upgraded to 5.3.1 (also, grgit version can be bumped to 3.1.1).

dejan2609 commented:
Related: https://github.com/scoverage/scalac-scoverage-plugin/issues/250 **_Add Scala 2.13.0-RC1 support_**

dejan2609 commented:
Update: **scalac-scoverage-plugin** and **scalac-scoverage-runtime** versions for 2.13.0-RC1 are released.

**scala-logging** release is pending:
- https://github.com/lightbend/scala-logging/pull/159 **_Update scala 2.13.0-M5 to 2.13.0-RC1_**
- https://repo1.maven.org/maven2/com/typesafe/scala-logging/

ijuma commented on README.md:
```diff
@@ -77,7 +77,7 @@ The release file can be found inside `./core/build/distributions/`.
     ./gradlew clean
 
-### Running a task with a particular version of Scala (either 2.11.x or 2.12.x) ###
+### Running a task with a particular *available* version of Scala (either 2.11.x, 2.12.x or 2.13.x) ###
```
Nit: I'd edit it to be:

"Running a task with one of the Scala versions available (2.11.x, 2.12.x or 2.13.x)"

dejan2609 replied:
Done.

ijuma commented on build.gradle:
```diff
@@ -423,6 +423,13 @@ subprojects {
+    if ( versions.baseScala.startsWith('2.13') ) {
+      scalaCompileOptions.additionalParameters.removeAll( [
+        "-Xlint:by-name-right-associative",
+        "-Xlint:unsound-match"
```
We should add a comment explaining why we are doing this.

dejan2609 replied:
Done.

ijuma commented on gradle/dependencies.gradle:
```diff
@@ -25,26 +25,37 @@ ext {
 // Add Scala version
 def defaultScala211Version = '2.11.12'
 def defaultScala212Version = '2.12.8'
+def availableScala213Version = '2.13.0-M5' // keep in mind to align milestone/rc Scala 2.13.0-XY version with 'scovarage' version below (and vice versa)
```
I think I'd just call this `defaultScala213Version`. It's the default for that release series. The fact that it's not available is orthogonal.

dejan2609 replied:
Done

ijuma commented on README.md:
```diff
@@ -86,7 +86,7 @@ You can pass either the major version (eg 2.12) or the full version (eg 2.12.7):
     ./gradlew -PscalaVersion=2.12 releaseTarGz
 
-### Running a task with all scala versions ###
+### Running a task with all scala *default* versions (2.11.x and 2.12.x) ###
```
Nit:

"Running a task with all the scala versions enabled by default (2.11.x and 2.12.x)"

dejan2609 replied:
Done.

ijuma reviewed (commented):
Thanks for the PR. I left some minor comments. Apart from that:

1. Please rebase as we just upgraded the Gradle version in trunk.
2. I think we can go straight to 2.13.0-RC1 since it seems like most libraries are already there.

Hopefully we can enable 2.13 by default for Kafka 2.4.

dejan2609 commented:
@ijuma done as requested (rebased onto trunk with suggested changes). 
Note: I will bump to Scala **2.13.0-RC1** as soon as **_scala-logging_** release appears.

ijuma commented on build.gradle:
```diff
@@ -422,6 +422,13 @@ subprojects {
+    if ( versions.baseScala.startsWith('2.13') ) {
+      scalaCompileOptions.additionalParameters.removeAll( [
+        "-Xlint:by-name-right-associative",
+        "-Xlint:unsound-match"
```
These are the two relevant PRs for these changes:

* https://github.com/scala/scala/pull/6502
* https://github.com/scala/scala/pull/5969

Instead of doing it like this, I'd do only add these params if baseScala is 2.11 or 2.12 (similar to `-Xlint:constant`). Future versions of Scala won't have them either so the suggested approach is more future proof. Also, worth including the PR references in a comment near the conditional code.

dejan2609 replied:
Done, new commit pushed.

ennru commented:
Did you consider replacing `scala-logging` with "manual" guards for the log level? AFAICS it's used in very few places and a dependency less might be better.

dejan2609 commented:
@ennru can't answer to your suggestion (out of my scope, at least for this particular PR).

@ijuma Scala **2.13.0-RC2** is available now:
- https://github.com/scala/scala/releases/tag/v2.13.0-RC2
- https://contributors.scala-lang.org/t/scala-2-13-0-rc2-release/3323

Note: there are some related `scala-logging` tickets, I presume that we will wait for an appropriate version.

dejan2609 commented:
`scala-logging` 3.9.2 is released 👍 but we also need `jackson-module-scala` compatible version, so I will squash and rebase onto trunk in order to pickup jackson version bump (2.9.8 -->> 2.9.9) performed by @ijuma few days ago.

**_jackson-module-scala_** related ticket: https://github.com/FasterXML/jackson-module-scala/issues/414

ijuma commented:
Thanks. What uses jackson-module-scala?

dejan2609 commented:
@ijuma  Actually, `jackson-module-scala` is referenced directly (via gradle/dependencies.gradle).

Squash, rebase and force-push are completed; another version bump (`scalatest` 3.0.7 -->> 3.0.8-RC4) is included.

Current build state (across different Scala versions) on my box (`./gradlew -PscalaVersion=2.1X jar`):
 * **2.11:** ❌ breaks with: `Could not find com.typesafe.scala-logging:scala-logging_2.11:3.9.2` -->> https://github.com/lightbend/scala-logging/issues/160
 * **2.12:** ✔ OK
 * **2.13:** ❌ breaks with: `Could not find com.fasterxml.jackson.module:jackson-module-scala_2.13.0-RC2:2.9.9` -->> https://github.com/FasterXML/jackson-module-scala/issues/414

dejan2609 commented:
**Update:** `scala-logging_2.11:3.9.2` has just been published (it means that Scala 2.11 build is OK now).

dejan2609 commented:
retest this please

dejan2609 commented:
**Update: Scala 2.13.0-RC3 is released:** 
- https://contributors.scala-lang.org/t/scala-2-13-0-rc2-release/3323/10
- https://github.com/scala/scala/releases/tag/v2.13.0-RC3

It seems that `jackson-module-scala` will release 2.9.9 versions for both RC2 and RC3, see this related comment: https://github.com/FasterXML/jackson-module-scala/pull/413#issuecomment-496835969

ijuma commented:
What depends on jackson-module-scala?

dejan2609 commented:
@ijuma ok, I get your question now: `jackson-module-scala` single usage that matters: 
**core** module / _**kafka.admin.ConsumerGroupCommand**_

Related PR: https://github.com/apache/kafka/pull/5726

ijuma commented:
@dejan2609 Thanks! I added a comment https://github.com/apache/kafka/pull/5726/files#r289078080

I think we should remove the Jackson Scala dependency personally. If you'd like to help with that, please reply to my comment above.

dejan2609 commented:
Another commit is pushed (scala RC3 version related):
- Scala: 2.13.0-RC2 -->> 2.13.0-RC3
- scalatest: 3.0.8-RC4 -->> 3.0.8-RC5
- scoverage: 1.4.0-RC2 -->> 1.4.0-RC3

Note: `scalaLogging` version (compiled against 2.13.0-RC3) is pending, but I guess it is just a matter of days: https://repo1.maven.org/maven2/com/typesafe/scala-logging (PR is created here: https://github.com/lightbend/scala-logging/pull/165, ~~but I need to print/sign/scan/email Lightbend Contributors License Agreement~~)

dejan2609 commented:
**scala-logging_2.13.0-RC3-3.9.2** is published (thanx @analytically); builds for Scala 2.11 and 2.12 are working fine, however 2.13 build returns error: `postfix operator headOption needs to be enabled
`
**Related:**
 * https://contributors.scala-lang.org/t/lets-drop-postfix-operators/1457/13
 * https://github.com/scala/scala/pull/6831 	
  
**Note:** @ijuma, @SethTisue **build.gradle** actually lists appropriate compiler option:

```
  tasks.withType(ScalaCompile) {
    scalaCompileOptions.additionalParameters = [
      "-deprecation",
      "-unchecked",
      "-encoding", "utf8",
      "-Xlog-reflective-calls",
      "-feature",
      "-language:postfixOps",
```
Scenario:	 
 * checkout this branch and execute **./gradlew -PscalaVersion=2.13 jar**
```
error: postfix operator headOption needs to be enabled
by making the implicit value scala.language.postfixOps visible.
This can be achieved by adding the import clause 'import scala.language.postfixOps'
or by setting the compiler option -language:postfixOps.
See the Scaladoc for value scala.language.postfixOps for a discussion
why the feature needs to be explicitly enabled.
          } headOption
            ^
one error found
```

ennru commented:
I'm not sure why the command line parameter doesn't do the trick. The best option is anyway to write the same line `}.headOption` with the explicit dot.

dejan2609 commented:
> 
> 
> I'm not sure why the command line parameter doesn't do the trick. The best option is anyway to write the same line `}.headOption` with the explicit dot.

@ennru and others:  I really need some help or at least some advice here. 
From way I see this, `headOption` is a Scala’s List collection class method and all I can use here is a Gradle DSL... Am I missing something ? :

dejan2609 commented:
I have a gut feeling that at least one of these is related; additional investigation is required and hence links are shown in plain text:
```
https://github.com/sbt/sbt/pull/4132   Avoid postfix op without feature import
https://github.com/scala/scala/pull/6559   Prepare to make postfixOps syntax an error (not just a warning) unless the feature is explicitly enabled
https://github.com/gradle/gradle/issues/2158   Scala compilation with sbt/zinc
https://github.com/gradle/gradle/pull/8485   upgraded Scala zinc compiler to version 1.2.5
https://github.com/gradle/gradle/pull/9573   Zinc version update
```

SethTisue commented:
@dejan2609 Enno isn't suggesting modifying the Gradle code, he's suggesting modifying the Scala code that is being compiled, changing this line:

```
          } headOption
```

to

```
          }.headOption
```

dejan2609 commented:
Thanx @SethTisue  ! I have a confession to make: as a build/release engineer I tend to cut corners and being in odds with Scala I am searching for a solutions elsewhere quite often/overzealously (although I mastered SBT's dark magic, just enough-to-be-mean).

Back to the subject: 
- bad news is that I can't change from `} headOption` to `}.headOption` that easy given a fact that error goes like this: 
```
Building project 'core' with Scala version 2.13.0-RC3
Building project 'streams-scala' with Scala version 2.13.0-RC3

> Task :core:compileScala
C:\Users\DEJANS\AppData\Local\Temp\sbt_5bb741e4\xsbt\Compat.scala:152: error: postfix operator headOption needs to be enabled
by making the implicit value scala.language.postfixOps visible.
This can be achieved by adding the import clause 'import scala.language.postfixOps'
or by setting the compiler option -language:postfixOps.
See the Scaladoc for value scala.language.postfixOps for a discussion
why the feature needs to be explicitly enabled.
          } headOption
            ^
one error found

> Task :core:compileScala FAILED
```
* **good news is that this is Windows only issue;** build on Linux ends up like this:

```
/home/dstojadinovic/kafka/core/src/main/scala/kafka/zk/ZkData.scala:46: Unused import
import scala.collection.{Seq, breakOut}
                              ^
/home/dstojadinovic/kafka/core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala:77: private val inFlightRequests in class ZooKeeperClient is never used
  private val inFlightRequests = new Semaphore(maxInFlightRequests)
              ^
608 warnings found
151 errors found

> Task :core:compileScala FAILED
```
I will file separate SBT ticket for this (although Scala devs on Windows boxes are hard to find, I guess).

dwijnand commented:
That was fixed almost 4 years ago: https://github.com/sbt/zinc/pull/5/commits/ecf121a7ae7ba99cddad69c91117948174dbfc15#diff-569dc93f697b5a143e904e517b998e1dR124.  I think you need to upgrade Kafka's Gradle setup to use a more recent version of Zinc.

dejan2609 commented:
@dwijnand Correct, I listed related gradle/zinc pulls/tickets above: https://github.com/apache/kafka/pull/5454#issuecomment-498856531. Links are in a plain text only (for the moment) because I dont have enough time to read them properly.

@ijuma is aware of this, I see his comments related to gradle/zinc version upgrade.

SethTisue commented:
> as a build/release engineer

understood, totally!

and anyway, as we've now seen, my suggestion was off-base anyway :-)

dejan2609 commented:
Rebased onto trunk and force-pushed in order to resolve conflicts in **gradle/dependencies.gradle** (after `scala-java8-compat` dependency is added).

ennru commented:
Seems like the Scala 2.11 and 2.12 builds are happy. So how does Scala 2.13 (which is published as final now) think about it?

dejan2609 commented:
Sure @ennru ! Scala **2.13.0** (i.e. final) build depends on these issues (dependencies versions):
- https://github.com/scalatest/scalatest/issues/1601
- https://github.com/scoverage/scalac-scoverage-plugin/issues/259 
- https://github.com/lightbend/scala-logging/issues/169
- https://github.com/FasterXML/jackson-module-scala/issues/416

Note: some Scala refactoring for 2.13 will have to take place.

ennru commented:
Scalatest 3.0.8 and scala-logging 3.9.2 are available for Scala 2.13.0 now. Getting there...

dejan2609 commented:
Ok, all dependencies are here, but I just digged some warnings that should be suppressed (`by-name-right-associative` and `unsound-match` made their way into Scala 2.13 build, I will have to investigate that).

dejan2609 commented:
Update: switch from `if ((versions.baseScala = '2.11') || (versions.baseScala = '2.12'))` to `if (versions.baseScala in ['2.11','2.12'])` did the trick (and it looks better, I guess).

Build works for Scala 2.11 / 2.12 and (unsurprisingly) returns some errors for Scala 2.13: **./gradlew clean -PscalaVersion=2.13 jar -q**

```
151 errors found

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':core:compileScala'.
> Compilation failed
```

ennru commented:
Great! 
So what would be the plan here: 
1) Will the build updates get merged and the code changes be worked on another PR, or
2) Should someone else branch this PR and suggest how to adapt to the Scala 2.13 (collection) APIs?

dejan2609 commented:
Thanx @ennru, I was just about to ask the same question.

@ijuma, @ewencp: what is your take on this ?

dejan2609 commented:
retest this please

dejan2609 commented:
Pinging @ijuma and @ewencp again.

ijuma commented:
@ennru We have to support Scala 2.11, 2.12 and 2.13. I don't expect an immediate code change to adapt to 2.13 collection APIs. Maybe further down the line when we drop 2.11 and consider switching to 2.13 as the recommended version (we'd still have to be compatible with 2.12 at that point).

ijuma commented:
retest this please

ijuma commented on build.gradle:
```diff
@@ -422,6 +420,15 @@ subprojects {
     }
+    
+// these options are valid for Scala versions < 2.13 only
+// Scala 2.13 removes them, see https://github.com/scala/scala/pull/6502 and https://github.com/scala/scala/pull/5969 
```
Nit: can we indent these comments properly please?

dejan2609 replied:
Done

ijuma commented on gradle/dependencies.gradle:
```diff
@@ -25,26 +25,38 @@ ext {
   // Available if -PscalaVersion is used. This is useful when we want to support a Scala version that has
   // a higher minimum Java requirement than Kafka. This was previously the case for Scala 2.12 and Java 7.
-  availableScalaVersions = [ '2.11', '2.12' ]
+  availableScalaVersions = [ '2.11', '2.12', '2.13' ]
```
Since 2.13 is out now, should we have this in the `defaultScalaVersions` instead?

dejan2609 replied:
Question and a note:

 **Question:** '_instead_' or '_also_' ? My assumption is that 2.13 should be included into `defaultScalaVersions` (but `availableScalaVersions` should keep version 2.13, right ?)

**Note:** if we add 2.13 into `defaultScalaVersions` gradle tasks **_testAll_**, **_jarAll_** and **_releaseTarGzAll_** will fail until code is refactored for 2.13 (but then again: maybe this is what you want).

ijuma replied:
I meant "also", sorry.

dejan2609 replied:
Done, 2.13 is added into `defaultScalaVersions`

dejan2609 commented:
@ijuma solved, rebased, squashed and force-pushed.

ijuma commented:
@dejan2609 Can you please rebase (unfortunately your branch includes a change that broke one test, it has since been fixed) and update the PR description to match the current reality? I should be able to merge then.

ijuma commented on README.md:
```diff
@@ -86,7 +86,7 @@ You can pass either the major version (eg 2.12) or the full version (eg 2.12.7):
     ./gradlew -PscalaVersion=2.12 releaseTarGz
 
-### Running a task with all scala versions ###
+### Running a task with all the scala versions enabled by default (2.11.x and 2.12.x) ###
```
Now that 2.13 is enabled by default, maybe we can remove the brackets.

dejan2609 replied:
Done, commit included/squashed.

ijuma commented on gradle/dependencies.gradle:
```diff
@@ -22,29 +22,41 @@ ext {
     versions["scala"] = scalaVersion
   }
 } else {
   versions["scala"] = defaultScala211Version
```
This should be `defaultScala212Version` since we default to 2.12 now. An existing issue, but maybe we can fix it at the same time.

dejan2609 replied:
Done, commit included/squashed.

ijuma reviewed (commented):
Sorry, noticed a couple of additional changes. We should be good to go after that.

ijuma reviewed (approved):
I made a couple of changes and it LGTM. It just needs a rebase and for the PR description to be updated.

dejan2609 commented:
@ijuma Done as requested, all changes are bundled into one commit. 
I took the liberty to alter PR description in a way that list both initial and final state (given a fact that PR exists for a while).

ijuma commented:
Scala 2.12 is green, Scala 2.11 had 1 unrelated flaky connect test. Merging to master.

ijuma commented:
While updating Jenkins to build with Scala 2.13, I ran into a number of compiler errors. I'll submit a PR with fixes.

ijuma commented:
See https://github.com/apache/kafka/pull/6989
